server/app/auth/oidc/applicant/OidcApplicantProfileAdapter.java

```diff
@@ -0,0 +1,127 @@
+package auth.oidc.applicant;
+
+import auth.CiviFormProfile;
+import auth.CiviFormProfileData;
+import auth.ProfileFactory;
+import auth.Roles;
+import auth.oidc.OidcProfileAdapter;
+import com.beust.jcommander.internal.Nullable;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.inject.Provider;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.oidc.client.OidcClient;
+import org.pac4j.oidc.config.OidcConfiguration;
+import org.pac4j.oidc.profile.OidcProfile;
+import repository.UserRepository;
+
+/**
+ * This class ensures that the OidcProfileCreator that both the AD and IDCS clients use will
+ * generate a CiviFormProfile object. This is necessary for merging those accounts with existing
+ * accounts - that's not usually needed in web applications which is why we have to write this class
+ * - pac4j doesn't come with it. It's abstract because AD and IDCS need slightly different
+ * implementations of the two abstract methods.
+ */
+public abstract class OidcApplicantProfileAdapter extends OidcProfileAdapter {
+  private final String emailAttributeName;
+  private final Optional<String> localeAttributeName;
+  private final ImmutableList<String> nameAttributeNames;
+
+  public OidcApplicantProfileAdapter(
+      OidcConfiguration configuration,
+      OidcClient client,
+      ProfileFactory profileFactory,
+      Provider<UserRepository> applicantRepositoryProvider,
+      String emailAttributeName,
+      @Nullable String localeAttributeName,
+      ImmutableList<String> nameAttributeNames) {
+    super(configuration, client, profileFactory, applicantRepositoryProvider);
+    this.emailAttributeName = Preconditions.checkNotNull(emailAttributeName);
+    this.localeAttributeName = Optional.ofNullable(localeAttributeName);
+    this.nameAttributeNames = Preconditions.checkNotNull(nameAttributeNames);
+  }
+
+  protected Optional<String> getName(OidcProfile oidcProfile) {
+    String name =
+        nameAttributeNames.stream()
+            .filter(s -> !s.isBlank())
+            .map((String attrName) -> oidcProfile.getAttribute(attrName, String.class))
+            .filter(s -> !Strings.isNullOrEmpty(s))
+            .collect(Collectors.joining(" "));
+
+    return Optional.ofNullable(Strings.emptyToNull(name));
+  }
+
+  protected Optional<String> getLocale(OidcProfile oidcProfile) {
+    return localeAttributeName
+        .map(name -> oidcProfile.getAttribute(name, String.class))
+        .filter(s -> !Strings.isNullOrEmpty(s));
+  }
+
+  // Legacy. TODO: remove.
```

Comment:
suggestion (blocking): file an issue for the TODO and describe why it's here and will be removed later